src/famfs_cli.c

```diff
@@ -1368,6 +1368,90 @@ do_famfs_cli_verify(int argc, char *argv[])
 	return 0;
 }
 
+/********************************************************************/
+void
+famfs_flush_usage(int   argc,
+	    char *argv[])
+{
+	char *progname = argv[0];
+
+	printf("\n"
+	       "famfs flush: Flush or invalidate the processor cache for an entire file\n"
+	       "\n"
+	       "This command is useful fur shared memory that is not cache coherent. It should\n"
+	       "be called after mutating a file whose mutations need to be visible on other hosts,\n"
+	       "and before accessing any file that may have been mutated on other hosts. Note that\n"
+	       "logplay also takes care of this, but if the log has not been played since the file\n"
+	       "was mutated, this operation may be needed.\n"
+	       "\n"
+	       "    %s flush [args] <file> [<file> ...]\n"
+	       "\n"
+	       "Arguments:\n"
+	       "    -v           - Verbose output\n"
+	       "    -?           - Print this message\n"
+	       "\nNOTE: this creates a file system error and is for testing only!!\n"
+	       "\n", progname);
+}
+
+int
+do_famfs_cli_flush(int argc, char *argv[])
+{
+	char fullpath[PATH_MAX];
+	char *file = NULL;
+	int verbose = 0;
+	int arg_ct = 0;
+	int errs = 0;
+	int rc;
+	int c;
+
+
+	/* XXX can't use any of the same strings as the global args! */
+	struct option cp_options[] = {
+		/* These options set a */
+		{0, 0, 0, 0}
+	};
+
+	/* Note: the "+" at the beginning of the arg string tells getopt_long
+	 * to return -1 when it sees something that is not recognized option
+	 * (e.g. the command that will mux us off to the command handlers
+	 */
+	while ((c = getopt_long(argc, argv, "+vh?",
+				cp_options, &optind)) != EOF) {
+
+		arg_ct++;
+		switch (c) {
+
+		case 'v':
+			verbose++;
+			break;
+		case 'h':
+		case '?':
+			famfs_flush_usage(argc, argv);
+			return 0;
+		}
+	}
+
+	if (optind > (argc - 1)) {
+		fprintf(stderr, "%s: source and destination filenames required\n", __func__);
+		famfs_clone_usage(argc, argv);
+		return -1;
+	}
+	while (optind < argc) {
+		file = argv[optind++];
+		if (realpath(file, fullpath) == NULL) {
+			fprintf(stderr, "%s: bad source path %s\n", __func__, file);
+			errs++;
```

Comment:
Should we break out or return when we hit the first error ? Because if there was an error it seems like we are going ahead and flushing the file ? (even if famfs_flush_file() handles invalid files, i think its better to error out here than to continue.... am i missing something ?!)

Reply:
Right, probably should "continue" rather than flushing bogus paths. There is a smoke test for this, and it does:

Smoke does the following:
smoke/test0.sh:${CLI} flush -?          || fail "flush -? should work"
smoke/test0.sh:${CLI} flush             && fail "flush with no args should fail"
smoke/test0.sh:${CLI} flush /bogus/file && "flush of a bogus file should fail"
smoke/test0.sh:${CLI} flush $(sudo find $MPT -type f -print) || fail "flush all files should work"
smoke/test0.sh:${CLI} flush -vv $(sudo find $MPT -print)     && fail "this flush should report errors"

So a bogus path was not causing a segfault when we flushed it. Adding "continue"